File: instaliker.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from urllib.request import Request, build_opener, HTTPCookieProcessor, HTTPHandler
from urllib.parse import urlencode
import http.cookiejar
import re
from json import loads

LOGIN = 'l0k9j8'
PASSWORD = '********'

EXCLUDED_USERS = []

INSTAGRAM_HOST = 'https://www.instagram.com'
AUTH_URI = '/accounts/login/ajax/'
LIKE_URI = '/web/likes/%s/like/'

HEADER = {'x-requested-with': 'XMLHttpRequest',
          'x-instagram-ajax': 1,
          'user-agent': 'Mozilla/5.0',
          'authority': 'www.instagram.com',
          'referer': 'https://www.instagram.com/'}


class ParseError(Exception):
    pass


class CoreError(Exception):
    pass


def csrf_token_from_cookie(cookies):
    for cookie in cookies:
        if cookie.name == 'csrftoken':
            return cookie.value
    return ''


def loader(opener, url, data=None, headers=None):
    req = Request(url, data=data, headers=headers)
    return opener.open(req).read().decode('utf-8')


def get_feed_page(user_feed):
    if 'entry_data' not in user_feed:
        raise ParseError('Entry data not found')
    if 'FeedPage' not in user_feed['entry_data']:
        raise ParseError('FeedPage not found')
    return user_feed['entry_data']['FeedPage']


def main():
    cookies = http.cookiejar.CookieJar()
    opener = build_opener(HTTPCookieProcessor(cookies), HTTPHandler())
    headers = HEADER.copy()
    loader(opener, INSTAGRAM_HOST, headers=headers)
    headers['x-csrftoken'] = csrf_token_from_cookie(cookies)
    auth_data = loads(loader(opener, INSTAGRAM_HOST + AUTH_URI, data=urlencode({'username': LOGIN,
                                                                                'password': PASSWORD}).encode('utf-8'),
                             headers=headers))

    if not auth_data['authenticated']:
        raise CoreError('Bad auth')
    feed_regexp = re.compile(r'window\._sharedData = (.+);</script>')
    feed_json_strings = feed_regexp.findall(loader(opener, INSTAGRAM_HOST, headers=headers))
    if feed_json_strings:
        user_feed_data = loads(feed_json_strings[0])
    else:
        raise CoreError('Bad regexp')

    headers['x-csrftoken'] = csrf_token_from_cookie(cookies)

    for feed in get_feed_page(user_feed_data):
        if 'feed' in feed and 'media' in feed['feed']:
            nodes = feed['feed']['media']['nodes']
        else:
            continue

        for node in nodes:
            print('Owner: %s, Image url: %s, Has liked: %s' % (node['owner']['username'],
                                                               node['display_src'],
                                                               node['likes']['viewer_has_liked']))
            if not node['likes']['viewer_has_liked'] and node['owner']['username'] not in EXCLUDED_USERS:
                loads(loader(opener, INSTAGRAM_HOST + LIKE_URI % node['id'], headers=headers, data=b''))
                print('Like image with id: %s' % node['id'])


if __name__ == '__main__':
    main()

File: instaliker.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"regexp"
	"runtime"
	"strings"
)

const INSTAGRAM_HOST string = "https://www.instagram.com"
const AUTH_URI string = "/accounts/login/ajax/"
const LIKE_URI string = "/web/likes/%s/like/"

type AuthJson struct {
	Status        string `json:"status"`
	Authenticated bool   `json:"authenticated"`
}

type LikeJson struct {
	Status string `json:"status"`
}

type InstaFeed struct {
	EntryData struct {
		Feedpage []struct {
			Feed struct {
				Media struct {
					Nodes []struct {
						Date    float64 `json:"date"`
						Caption string  `json:"caption"`
						Likes   struct {
							ViewerHasLiked bool `json:"viewer_has_liked"`
						} `json:"likes"`
						Owner struct {
							Username string `json:"username"`
						} `json:"owner"`
						ID         string `json:"id"`
						DisplaySrc string `json:"display_src"`
					} `json:"nodes"`
				} `json:"media"`
			} `json:"feed"`
			Suggesteduserslist interface{} `json:"suggestedUsersList"`
		} `json:"FeedPage"`
	} `json:"entry_data"`
}

func loader(client *http.Client, urlAddr string, header http.Header, data *url.Values) []byte {
	method := "GET"
	if data != nil {
		method = "POST"
	} else {
		data = &url.Values{}
	}
	request, _ := http.NewRequest(method, urlAddr, bytes.NewBufferString(data.Encode()))
	request.Header = header
	resp, _ := client.Do(request)
	defer resp.Body.Close()
	content, _ := ioutil.ReadAll(resp.Body)
	return content
}

func csrfTokenFromCookies(cookiejar *cookiejar.Jar) string {
	urlParse, _ := url.Parse(INSTAGRAM_HOST)
	for _, cookie := range cookiejar.Cookies(urlParse) {
		if cookie.Name == "csrftoken" {
			return cookie.Value
		}
	}
	return ""
}

func instaliker(login *string, password *string, users map[string]bool, custom_users bool) {
	var header = http.Header{
		"x-requested-with": []string{"XMLHttpRequest"},
		"x-instagram-ajax": []string{"1"},
		"user-agent":       []string{"Mozilla/5.0"},
		"authority":        []string{"www.instagram.com"},
		"referer":          []string{"https://www.instagram.com/"},
	}
	jar, _ := cookiejar.New(nil)
	client := &http.Client{Jar: jar}
	loader(client, INSTAGRAM_HOST, header, nil)
	header.Set("x-csrftoken", csrfTokenFromCookies(jar))
	data := &url.Values{}
	data.Add("username", *login)
	data.Add("password", *password)
	var auth_resp *AuthJson
	json.Unmarshal(loader(client, INSTAGRAM_HOST+AUTH_URI, header, data), &auth_resp)
	if !auth_resp.Authenticated {
		fmt.Println("Access denied!")
		return
	}
	header.Set("x-csrftoken", csrfTokenFromCookies(jar))
	re := regexp.MustCompile("window\\._sharedData = (.+);</script>")
	data_str := re.FindSubmatch(loader(client, INSTAGRAM_HOST, header, data))
	if len(data_str) < 2 {
		fmt.Println("Data parse error!")
		return
	}
	var insta_feed *InstaFeed
	json.Unmarshal(data_str[1], &insta_feed)
	header.Set("x-csrftoken", csrfTokenFromCookies(jar))
	for _, feed := range insta_feed.EntryData.Feedpage {
		done_ch := make(chan bool, len(feed.Feed.Media.Nodes))
		length := 0
		for _, node := range feed.Feed.Media.Nodes {
			like_it, ok := users[node.Owner.Username]
			if custom_users {
				like_it = ok && like_it
			} else {
				like_it = !ok || like_it
			}
			if !node.Likes.ViewerHasLiked && like_it {
				fmt.Println(fmt.Sprintf("Like user %s, Image: \"%s\" URL: %s", node.Owner.Username,
					node.Caption,
					node.DisplaySrc))
				length = length + 1
				go func(done chan<- bool, urlLike string) {
					loader(client, urlLike, header, &url.Values{})
					done <- true
				}(done_ch, fmt.Sprintf(INSTAGRAM_HOST+LIKE_URI, node.ID))
			}
		}
		for i := 0; i < length; i = i + 1 {
			<-done_ch
		}
	}
}

func main() {
	runtime.GOMAXPROCS(runtime.NumCPU())
	login := flag.String("login", "", "instagram login")
	password := flag.String("password", "", "instagram password")
	exclude_users_string := flag.String("exclude", "", "exclude from autolike (comma-separated list)")
	custom_users_string := flag.String("users", "", "autolike only for custom users (comma-separated list)")
	flag.Parse()
	if *login == "" {
		fmt.Printf("Instagram login not found")
		return
	}
	if *password == "" {
		fmt.Printf("Instagram password not found")
		return
	}
	users := make(map[string]bool)
	if (*exclude_users_string != "") && (*custom_users_string != "") {
		fmt.Printf("Use exclude or users")
		return
	}

	for _, excl_user := range strings.Split(*exclude_users_string, ",") {
		if excl_user != "" {
			users[excl_user] = false
		}
	}
	for _, incl_user := range strings.Split(*custom_users_string, ",") {
		if incl_user != "" {
			users[incl_user] = true
		}
	}
	instaliker(login, password, users, *custom_users_string != "")
}
